lib/lexer/lexer.cpp

```diff
@@ -0,0 +1,484 @@
+Backslash -> push state + switch state
+(D|S)Quote/Expansion/BackQuote -> pushChar + consume + pushState + switch state
+Space || EOF -> delimit + switch to Empty
+Comment -> delimit + switch to Comment
+*/
+Lexer::State Lexer::wordState() {
+	State state = State::Word;
+	while (state == State::Word) {
+		char ch = chars.peek();
+
+		// [ push char, consume, push state ], switch state
+		if (isSingleQuote(ch) || isDoubleQuote(ch) || isDollarSign(ch) || isBackQuote(ch)) {
+			state_data.word.push_back(chars.consume());
+			state_data.pushState(State::Word);
+			if (isSingleQuote(ch)) {
+				return State::SingleQuote;
+			} else if (isDoubleQuote(ch)) {
+				return State::DoubleQuote;
+			} else if (isDollarSign(ch)) {
+				return State::ExpansionStart;
+			} else if (isBackQuote(ch)) {
+				return State::BackQuote;
+			}
+		}
+		// [ push state, switch state ]
+		if (isBackslash(ch)) {
+			state_data.pushState(State::Word);
+			return State::Backslash;
+		}
+
+		// [ delimit ], switch state
+		if (isSpace(ch) || ch == EOF) {
+			state = State::Empty;
+		} else if (isCommentStart(ch)) {
+			state = State::Comment;
+		} else if (isOperatorStart(ch)) {
+			state = State::Operator;
+		} else {
+			// Continue word state
+			state_data.word.push_back(chars.consume());
+		}
+	}
+	// Handle delimit outside of loop: EOF, IFS, operator, comment
+	if (!state_data.word.empty()) {
+		if (isIoNumber(state_data.word, chars.peek())) {
+			delimit(Token(Token::Type::IoNumber, IoNumber{std::move(state_data.word)}));
+		} else {
+			delimit(Token{Token::Type::Token, WordToken{std::move(state_data.word)}});
+		}
+		state_data.word.clear();
+	}
+	return state;
+}
+
+Lexer::State Lexer::operatorState() {
+	char ch = chars.peek();
+
+	if (isBackslash(ch)) {
+		state_data.pushState(State::Operator);
+		return State::Backslash;
+	}
+
+	int matches;
+	if (!isOperatorCharacter(ch)) {
+		matches = 0;
+	} else {
+		matches = Token::prefixOperatorMatches(state_data.word + ch);
+	}
+
+	if (matches >= 1) {
+		// The characters we have now are part of 1 or multiple operators
+		// don't delimit yet, because this does not tell us that it matches an entire operator
+		state_data.word.push_back(ch);
+		chars.consume();
+		return State::Operator;
+	}
+	delimitOperator();
+	return State::Empty;
+}
+
+// Previously it did match a prefix of a valid operator
+// but with the additional character it doesn't;
+// - find out what the previous operator was
+void Lexer::delimitOperator() {
+	auto result = Token::exactOperatorType(state_data.word);
+	D_ASSERT(result.has_value());
+	delimit(Token {result.value(), OperatorToken()});
+	state_data.word.clear();
+}
+
+Lexer::State Lexer::commentState() {
+	char ch = chars.peek();
+	if (isNewline(ch) || ch == EOF) {
+		return State::Empty;
+	}
+	chars.consume();
+	return State::Comment;
+}
+
+// Precondition: backslash is NOT consumed. Previous state is either WORD or OPERATOR
+Lexer::State Lexer::backslashState() {
+	char ch = chars.consume();
```

Comment:
This does not add `\` to the produced word right?
I think that's not according to the spec
https://pubs.opengroup.org/onlinepubs/009695399/utilities/xcu_chap02.html#tag_02_06_07

Quote Removal is done in a later stage (after the initial 3 stages of expansion are applied generally)

Reply:
It adds `\` at the end of the function, only if the next character is not a <newline>, who then must be removed from the input stream together with the backslash:
https://pubs.opengroup.org/onlinepubs/009695399/utilities/xcu_chap02.html#tag_02_02_01

If a <newline> follows the backslash, the shell shall interpret this as line continuation. The backslash and <newline>s shall be removed before splitting the input into tokens.

Reply:
This sounds kind of confusing though:
> If a <newline> follows the backslash, the shell shall interpret this as line continuation. The backslash and <newline>s shall be removed before splitting the input into tokens.

(from https://pubs.opengroup.org/onlinepubs/009695399/utilities/xcu_chap02.html#tag_02_02_01)

Reply:
But it is mentioned in the section of Quoting (https://pubs.opengroup.org/onlinepubs/009695399/utilities/xcu_chap02.html#tag_02_02)

> The various quoting mechanisms are the escape character, single-quotes, and double-quotes

Reply:
Ah.. I can't read
You manually add the backslash after you've dealt with the cases where it should be ignored (newline)